rs/rosetta-api/icrc1/ledger/tests/tests.rs

```diff
@@ -1543,3 +1553,151 @@ mod verify_written_blocks {
+                node_max_memory_size_bytes: None,
+                max_message_size_bytes: None,
+                controller_id: PrincipalId::new_user_test_id(100),
+                more_controller_ids: None,
+                cycles_for_archive_creation: None,
+                max_transactions_per_response: None,
+            },
+            max_memo_length: None,
+            feature_flags: Some(FeatureFlags { icrc2: false }),
+            maximum_number_of_accounts: None,
+            accounts_overflow_trim_quantity: None,
+        }))
+        .unwrap()
+    }
+
+    #[test]
+    fn should_successfully_upgrade_ledger_from_u64_to_u256_to_u64_wasm() {
+        let env = StateMachine::new();
+        let ledger_id = env
+            .install_canister(ledger_mainnet_u64_wasm(), default_init_args(), None)
+            .unwrap();
+        // Create a large balance that is still smaller than the max token supply of u64::MAX
+        transfer(&env, ledger_id, MINTER, account(1), u64::MAX - 100);
+        let mut balance_1 = balance_of(&env, ledger_id, account(1));
+        // Create a large allowance that gets capped at u64::MAX
+        let mut big_amount = BigUint::ZERO;
+        big_amount.set_bit(65, true);
+        let approval_result = send_approval(
+            &env,
+            ledger_id,
+            account(1).owner,
+            &ApproveArgs {
+                from_subaccount: None,
+                spender: account(2),
+                amount: Nat::from(big_amount),
+                expected_allowance: None,
+                expires_at: None,
+                fee: None,
+                memo: None,
+                created_at_time: None,
+            },
+        );
+        assert_eq!(approval_result, Ok(BlockIndex::from(1u64)));
+        balance_1 -= FEE;
+
+        // Try to upgrade the ledger from using a u64 wasm to a u256 wasm
+        let upgrade_args = Encode!(&LedgerArgument::Upgrade(None)).unwrap();
+        env.upgrade_canister(ledger_id, ledger_mainnet_u256_wasm(), upgrade_args)
+            .expect("Unable to upgrade the ledger canister");
+
+        // The balance and allowance should not change
+        assert_eq!(balance_1, balance_of(&env, ledger_id, account(1)));
+        let allowance_2 = get_allowance(&env, ledger_id, account(1), account(2));
+        assert_eq!(allowance_2.allowance, Nat::from(u64::MAX));
+
+        // Try to upgrade the ledger back to a u64 wasm
+        let upgrade_args = Encode!(&LedgerArgument::Upgrade(None)).unwrap();
+        env.upgrade_canister(ledger_id, ledger_mainnet_u64_wasm(), upgrade_args)
+            .expect("Unable to upgrade the ledger canister");
+
+        // The balance and allowance should not change
+        assert_eq!(balance_1, balance_of(&env, ledger_id, account(1)));
+        assert_eq!(allowance_2.allowance, Nat::from(u64::MAX));
+    }
+
+    #[test]
+    fn should_fail_to_upgrade_to_u256_ledger_wasm_if_balance_too_large() {
```

Comment:
nit: isn't it the total supply that's too large here and not the balance?

```suggestion
    fn should_fail_to_upgrade_to_u256_ledger_wasm_if_total_supply_too_large() {
```

Reply:
Yes! I also changed `u256` to `u64` in the test name since that's the upgrade path that fails.